### PR TITLE
feat: restricts unverified user for buying free tickets

### PR DESCRIPTION
**Pull request by shreyanshdwivedi**
Related to https://github.com/fossasia/open-event-android/issues/1881#issuecomment-499376863

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Changes proposed in this pull request:
It adds a check to restrict an unverified user from buying free tickets

### CODE REVIEW

**Comment by kushthedude** [comment]
@shreyanshdwivedi Are you sure all mentors have approved it because this will break the flow for Guest Order Form ?

**Comment by shreyanshdwivedi** [comment]
@kushthedude please refer to the issue mentioned and @iamareebjamal asked me to add this check on server later in that issue itself.
Also, I've checked for guest orders, it'll only stop them from buying free tickets. Paid tickets can be bought. Please check and review

**Comment by shreyanshdwivedi** [comment]
@mrsaicharan1 @iamareebjamal @uds5501 @kushthedude please review

File: app/api/attendees.py [hunk 1]
```diff
@@ -170,6 +170,17 @@ def before_update_object(self, obj, data, kwargs):
 #         if not has_access('is_registrar', event_id=obj.event_id):
 #         raise ForbiddenException({'source': 'User'}, 'You are not authorized to access this.')
 
+        if 'ticket' in data:
+            user = safe_query(self, User, 'id', current_identity.id, 'user_id')
+            ticket = db.session.query(Ticket).filter_by(
+                id=int(data['ticket']), deleted_at=None
+                ).first()
+            if not (user.is_verified or ticket.price):
```
Comment (iamareebjamal) [inline app/api/attendees.py]: Condition should be not user.is_verified and ticket.price == 0

It should be clear

File: app/api/attendees.py [hunk 2]
```diff
@@ -170,6 +170,17 @@ def before_update_object(self, obj, data, kwargs):
 #         if not has_access('is_registrar', event_id=obj.event_id):
 #         raise ForbiddenException({'source': 'User'}, 'You are not authorized to access this.')
 
+        if 'ticket' in data:
+            user = safe_query(self, User, 'id', current_identity.id, 'user_id')
+            ticket = db.session.query(Ticket).filter_by(
+                id=int(data['ticket']), deleted_at=None
+                ).first()
```
Comment (iamareebjamal) [inline app/api/attendees.py]: Handle the case if ticket is not found

**Comment by shreyanshdwivedi** [comment]
@uds5501 @iamareebjamal I've updated the checks. Please review

File: app/api/attendees.py [hunk 3]
```diff
@@ -170,6 +170,21 @@ def before_update_object(self, obj, data, kwargs):
...
+            if ticket is None:
+                raise UnprocessableEntity(
+                    {'pointer': '/data/relationships/ticket'}, "Invalid Ticket"
+                )
+            if not user.is_verified and ticket.price == 0:
```
Comment (uds5501) [inline app/api/attendees.py]: @shreyanshdwivedi just want to ask, what will happen in the case where a guest user decides to buy a ticket. AFAIK, they will be stopped from buying any ticket if they are in combination with a free order too?

Comment (shreyanshdwivedi) [reply app/api/attendees.py]: Yes @uds5501. When an order is made which have multiple tickets, at first a PATCH request is sent to add the info of the attendees and then order object is created.
But we are raising error at the addition of attendees itself.
I've tested a combination of paid and free tickets and error was raised. When I checked orders, no order object was created.
However please test it on local and confirm if this implementation is working for u or not.

**Comment by kushthedude** [comment]
@shreyanshdwivedi , Why did we restricted the orders of combination of Paid and Free Tickets for unverified users?

**Comment by shreyanshdwivedi** [comment]
@kushthedude beacuse the issue I mentioned asked to apply check on free tickets - not free orders

**Comment by kushthedude** [comment]
@shreyanshdwivedi @iamareebjamal Shouldn't we restrict free orders rather than free tickets ?

**Pull request merged**